### PR TITLE
Fix #269467: added support to ignore diacritic accents

### PR DESCRIPTION
**Pull request by finleylau**
Used regex to replace special characters with accents, acutes, grave, umlaut, etc. with corresponding English characters

### CODE REVIEW

File: mscore/instrwidget.cpp [hunk 1]
```diff
@@ -56,6 +56,26 @@ void filterInstruments(QTreeWidget* instrumentList, const QString &searchPhrase)
...
+                  text = text.replace(QRegularExpression("[\\x{00C1}\\x{00C2}\\x{00C4}\\x{00C0}\\x{00C3}]"), QChar('a'));
+                  text = text.replace(QRegularExpression("[\\x{00C9}\\x{00CA}\\x{00CB}\\x{00C8}]"), QChar('e'));
+                  text = text.replace(QRegularExpression("[\\x{00CD}\\x{00CE}\\x{00CF}\\x{00CC}]"), QChar('i'));
+                  text = text.replace(QRegularExpression("[\\x{00D3}\\x{00D4}\\x{00D6}\\x{00D2}\\x{00D5}]"), QChar('o'));
+                  text = text.replace(QRegularExpression("[\\x{00DA}\\x{00DB}\\x{00DC}\\x{00D9}]"), QChar('u'));
+                  text = text.replace(QRegularExpression("[\\x{00C7}]"), QChar('c'));
+                  text = text.replace(QRegularExpression("[\\x{00D1}]"), QChar('n'));
+                  text = text.replace(QRegularExpression("[\\x{00DD}]"), QChar('y'));
+
                   bool isMatch = text.contains(searchPhrase, Qt::CaseInsensitive);
```
Comment (Jojo-Schmitz) [inline mscore/instrwidget.cpp]: Instead you could directly use a `contains()` variant that works with regular expressions, like `bool QString::contains(const QRegularExpression &re) const`.
Also what about e.g. the Dutch `ij` and the German `ß`. Which characters are you covering, can you add a comment about that, e.g. mentioning their HTML entity, like `&auml;` for `ä`?

**Comment by lasconic** [comment]
I like this approach better https://stackoverflow.com/questions/12278448/removing-accents-from-a-qstring#12806471 but I don't know how it would behave for CJK, russian, arabic, hebrew and others.

**Comment by finleylau** [comment]
For non-Latin languages, is the goal to allow them to type Latin characters and have for example Russian characters appear?  or is the goal to find a way to replace accented Latin characters with unaccented ones, but still leave the non-Latin characters the same?

**Comment by lasconic** [comment]
I would leave non Latin languages alone, I assume someone using MuseScore in russian can type russian and doesn't need transliteration.

**Comment by finleylau** [comment]
I've implemented it currently to the first solution suggested in https://stackoverflow.com/questions/12278448/removing-accents-from-a-qstring#12806471.  The answer posted replaces too many characters and potentially could mess up many non Latin languages.  The code is now decomposing the Unicode, making sure that each Unicode character is not an accent, and then recomposing the Unicode value.  However, it doesn't work for the ß or æ or similar characters, as  these characters are considered by the QT Unicode system as being "Lower Case Letters" and therefore can't programatically be decomposed and recomposed like other accented characters.

One solution would be to leave them in original form and allow users to just type the nonLatin characters and leave the solution as it stands now.  Another more complicated solution would be to manually replace characters like Ø and the above mentioned characters and allow for both typing the nonLatin characters OR typing the corresponding Latin characters.  However, I'm not aware of the correct Latin spelling, for example I've read that ß can be represented by either 'ss', 'sz', or just 's'.  Do you have any suggestions?

File: mscore/instrwidget.cpp [hunk 2]
```diff
@@ -56,7 +56,45 @@ void filterInstruments(QTreeWidget* instrumentList, const QString &searchPhrase)
...
+                                    // eszett, &#223
+                                    manualReplaceDiacritic.append(QString("ss"));
+                                    }
+                              else if (text.at(i) == QChar(0x0133)) {
+                                    // ij, &#306
+                                    manualReplaceDiacritic.append(QString("ij"));
+                                    }
+                              else if (text.at(i) == QChar(0x00E6)) {
+                                    // ae, &#230
+                                    manualReplaceDiacritic.append(QString("ae"));
```
Comment (Jojo-Schmitz) [inline mscore/instrwidget.cpp]: There's also äüöÄÜÖ, transcribed as ae, oe. ue, Ae, Oe, Ue, and also there's œ, Œ, and Æ, transcribed as oe, OE, AE
There may be more

File: mscore/instrwidget.cpp [hunk 3]
```diff
@@ -56,7 +56,45 @@ void filterInstruments(QTreeWidget* instrumentList, const QString &searchPhrase)
...
+                  QString manualReplaceDiacritic("");
+                  if (text.contains(QRegularExpression("[\\x{00F8}\\x{00DF}\\x{0133}\\x{00E6}]"))) {
+                        for (int i = 0; i < text.length(); ++i) {
+                              if (text.at(i) == QChar(0x00F8)) {
+                                    // o with stroke, &#248
+                                    manualReplaceDiacritic.append(QChar('o'));
+                                    }
+                              else if (text.at(i) == QChar(0x00DF)) {
+                                    // eszett, &#223
+                                    manualReplaceDiacritic.append(QString("ss"));
```
Comment (Jojo-Schmitz) [inline mscore/instrwidget.cpp]: indeed the correct transcription, even if it gets pronounced as "sz" or "sharp s"

**Comment by lasconic** [comment]
*should i do them in mtest/scripting/tst_scripting.cpp  or should i make a tst stringutils project in libmscore*

I would create a new mtest/stringutils/tst_stringutils.cpp file

**Comment by finleylau** [comment]
Added the stringutils class and added three tests (normal string that shouldn't be modified, one to replace all the ligatures, and one to replace the diacritics)

**Comment by Jojo-Schmitz** [comment]
That merge commit shouldn't be there, don't merge, but rebase

**Comment by finleylau** [comment]
Continued in #3496

**Pull request closed without merging**